src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -984,6 +984,253 @@ private Task WriteSuffix()
+            var percentageIndex = pathToEndSpan.IndexOf(BytePercentage);
+
+            var pathEnd = MinNonZero(spaceIndex, questionMarkIndex, percentageIndex);
+
+            if (spaceIndex == -1 && questionMarkIndex == -1 && percentageIndex == -1)
+            {
+                RejectRequestLine(start, end);
+            }
+            else if (percentageIndex != -1)
+            {
+                needDecode = true;
+
+                pathEnd = MinNonZero(spaceIndex, questionMarkIndex);
+                if (questionMarkIndex == -1 && spaceIndex == -1)
+                {
+                    RejectRequestLine(start, end);
+                }
+            }
+
+            var queryString = "";
+            var queryEnd = pathEnd;
+            if (questionMarkIndex != -1)
+            {
+                if (spaceIndex == -1)
+                {
+                    RejectRequestLine(start, end);
+                }
+
+                queryString = pathToEndSpan.Slice(pathEnd, queryEnd).GetAsciiString();
+            }
+
+            if (pathBegin == pathEnd)
+            {
+                RejectRequestLine(start, end);
+            }
+
+            var versionBegin = queryEnd + 1;
+            var versionToEndSpan = pathToEndSpan.Slice(versionBegin, pathToEndSpan.Length - versionBegin);
+            // var ves = Encoding.ASCII.GetString(versionToEndSpan.ToArray());
+            versionBegin = 0;
+            var versionEnd = versionToEndSpan.IndexOf(ByteCR);
+
+            if (versionEnd == -1)
+            {
+                RejectRequestLine(start, end);
+            }
+
+            if (!versionToEndSpan.GetKnownVersion(out string httpVersion))
+            {
+                httpVersion = new Utf8String(versionToEndSpan.Slice(0, versionEnd)).ToString();
+
+                if (httpVersion == string.Empty)
+                {
+                    RejectRequestLine(start, end);
+                }
+                else
+                {
+                    RejectRequest(RequestRejectionReason.UnrecognizedHTTPVersion, httpVersion);
+                }
+            }
+
+            if (versionToEndSpan[versionToEndSpan.Length - 1] != ByteLF)
+            {
+                RejectRequestLine(start, end);
+            }
+
+            var pathBuffer = pathToEndSpan.Slice(pathBegin, pathEnd);
+            var targetBuffer = pathToEndSpan.Slice(pathBegin, queryEnd);
+
+            // URIs are always encoded/escaped to ASCII https://tools.ietf.org/html/rfc3986#page-11
+            // Multibyte Internationalized Resource Identifiers (IRIs) are first converted to utf8;
+            // then encoded/escaped to ASCII  https://www.ietf.org/rfc/rfc3987.txt "Mapping of IRIs to URIs"
+            string requestUrlPath;
+            string rawTarget;
+            if (needDecode)
+            {
+                // Read raw target before mutating memory.
+                rawTarget = targetBuffer.GetAsciiString() ?? string.Empty;
```

Comment:
This should never be null. The target must either be in absolute form, or start with a `/`. cc @natemcmaster who looked into this part of the spec not long ago.